File: functionsystem/src/runtime_manager/executor/sandbox/runtime_state_manager.cpp
/*
 * Copyright (c) Huawei Technologies Co., Ltd. 2025. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "runtime_state_manager.h"

#include "common/logs/logging.h"

namespace functionsystem::runtime_manager {

// ── Registration ──────────────────────────────────────────────────────────────

void RuntimeStateManager::Register(SandboxInfo info)
{
    const std::string &runtimeID = info.runtimeID;
    YRLOG_DEBUG("RuntimeStateManager::Register runtimeID({})", runtimeID);
    sandboxes_[runtimeID] = std::move(info);
}

void RuntimeStateManager::Unregister(const std::string &runtimeID)
{
    YRLOG_DEBUG("RuntimeStateManager::Unregister runtimeID({})", runtimeID);
    sandboxes_.erase(runtimeID);
    inProgressStarts_.erase(runtimeID);
    pendingDeletes_.erase(runtimeID);
    // Note: warmUpMap_ is intentionally NOT cleared here.
    // WarmUp state has its own lifecycle via RegisterWarmUp/UnregisterWarmUp.
}

// ── Queries ───────────────────────────────────────────────────────────────────

std::optional<SandboxInfo> RuntimeStateManager::Find(const std::string &runtimeID) const
{
    auto it = sandboxes_.find(runtimeID);
    if (it == sandboxes_.end()) {
        return std::nullopt;
    }
    return it->second;
}

bool RuntimeStateManager::IsActive(const std::string &runtimeID) const
{
    return sandboxes_.count(runtimeID) > 0;
}

bool RuntimeStateManager::HasSandbox(const std::string &runtimeID) const
{
    return sandboxes_.count(runtimeID) > 0;
}

std::string RuntimeStateManager::GetSandboxID(const std::string &runtimeID) const
{
    auto it = sandboxes_.find(runtimeID);
    return it != sandboxes_.end() ? it->second.sandboxID : std::string{};
}

std::string RuntimeStateManager::GetCheckpointID(const std::string &runtimeID) const
{
    auto it = sandboxes_.find(runtimeID);
    return it != sandboxes_.end() ? it->second.checkpointID : std::string{};
}

void RuntimeStateManager::SetCheckpointID(const std::string &runtimeID, const std::string &checkpointID)
{
    auto it = sandboxes_.find(runtimeID);
    if (it == sandboxes_.end()) {
        YRLOG_WARN("RuntimeStateManager::SetCheckpointID runtimeID({}) not found", runtimeID);
        return;
    }
    it->second.checkpointID = checkpointID;
}

void RuntimeStateManager::ClearCheckpointID(const std::string &runtimeID)
{
    auto it = sandboxes_.find(runtimeID);
    if (it != sandboxes_.end()) {
        it->second.checkpointID.clear();
    }
}

std::map<std::string, messages::RuntimeInstanceInfo> RuntimeStateManager::GetAllInstanceInfos() const
{
    std::map<std::string, messages::RuntimeInstanceInfo> result;
    for (const auto &[id, info] : sandboxes_) {
        result[id] = info.instanceInfo;
    }
    return result;
}

std::unordered_map<std::string, SandboxInfo> RuntimeStateManager::GetAllSandboxes() const
{
    return sandboxes_;
}

std::string RuntimeStateManager::GetPortMappingsJson(const std::string &runtimeID) const
{
    auto it = sandboxes_.find(runtimeID);
    if (it == sandboxes_.end()) {
        return "";
    }
    return it->second.portMappingsJson;
}

// ── Partial updates ───────────────────────────────────────────────────────────

void RuntimeStateManager::UpdateSandboxID(const std::string &runtimeID, const std::string &sandboxID)
{
    auto it = sandboxes_.find(runtimeID);
    if (it == sandboxes_.end()) {
        YRLOG_WARN("RuntimeStateManager::UpdateSandboxID runtimeID({}) not found", runtimeID);
        return;
    }
    it->second.sandboxID = sandboxID;
}

void RuntimeStateManager::UpdateCheckpoint(const std::string &runtimeID, const std::string &checkpointID)
{
    auto it = sandboxes_.find(runtimeID);
    if (it == sandboxes_.end()) {
        YRLOG_WARN("RuntimeStateManager::UpdateCheckpoint runtimeID({}) not found", runtimeID);
        return;
    }
    it->second.checkpointID = checkpointID;
}

void RuntimeStateManager::UpdatePortMappings(const std::string &runtimeID, const std::string &portMappingsJson)
{
    auto it = sandboxes_.find(runtimeID);
    if (it == sandboxes_.end()) {
        YRLOG_WARN("RuntimeStateManager::UpdatePortMappings runtimeID({}) not found", runtimeID);
        return;
    }
    it->second.portMappingsJson = portMappingsJson;
}

// ── In-progress start tracking ────────────────────────────────────────────────

void RuntimeStateManager::MarkStartInProgress(const std::string &runtimeID,
                                               litebus::Future<messages::StartInstanceResponse> future)
{
    inProgressStarts_.emplace(runtimeID, std::move(future));
}

void RuntimeStateManager::MarkStartDone(const std::string &runtimeID)
{
    inProgressStarts_.erase(runtimeID);
}

bool RuntimeStateManager::IsStartInProgress(const std::string &runtimeID) const
{
    return inProgressStarts_.count(runtimeID) > 0;
}

std::optional<litebus::Future<messages::StartInstanceResponse>> RuntimeStateManager::GetInProgressFuture(
    const std::string &runtimeID) const
{
    auto it = inProgressStarts_.find(runtimeID);
    if (it == inProgressStarts_.end()) {
        return std::nullopt;
    }
    return it->second;
}

// ── Pending-delete tracking ───────────────────────────────────────────────────

void RuntimeStateManager::MarkPendingDelete(const std::string &runtimeID)
{
    pendingDeletes_.insert(runtimeID);
}

void RuntimeStateManager::ClearPendingDelete(const std::string &runtimeID)
{
    pendingDeletes_.erase(runtimeID);
}

bool RuntimeStateManager::IsPendingDelete(const std::string &runtimeID) const
{
    return pendingDeletes_.count(runtimeID) > 0;
}

// ── Warm-up state ─────────────────────────────────────────────────────────────

void RuntimeStateManager::RegisterWarmUp(const std::string &runtimeID, runtime::v1::FunctionRuntime proto)
{
    warmUpMap_[runtimeID] = std::move(proto);
}

void RuntimeStateManager::UnregisterWarmUp(const std::string &runtimeID)
{
    warmUpMap_.erase(runtimeID);
}

bool RuntimeStateManager::IsWarmUp(const std::string &runtimeID) const
{
    return warmUpMap_.count(runtimeID) > 0;
}

std::optional<runtime::v1::FunctionRuntime> RuntimeStateManager::GetWarmUp(const std::string &runtimeID) const
{
    auto it = warmUpMap_.find(runtimeID);
    if (it == warmUpMap_.end()) {
        return std::nullopt;
    }
    return it->second;
}

}  // namespace functionsystem::runtime_manager

File: functionsystem/src/runtime_manager/src/sandbox/runtime_state_manager.rs
//! Aggregated sandbox state — port of C++ `runtime_manager/executor/sandbox/
//! runtime_state_manager.{h,cpp}`.
//!
//! Owns and enforces sandbox-state invariants behind a single lock so
//! `register`/`unregister` are atomic across all internal maps (the C++ code got
//! this from its single actor thread; in Rust we use one `Mutex<Inner>`).
//! WarmUp state and active-sandbox state are orthogonal: a runtime_id is in at
//! most one of them.

use std::collections::{HashMap, HashSet};

use parking_lot::Mutex;
use yr_proto::messages::RuntimeInstanceInfo;
use yr_proto::runtime::v1::FunctionRuntime;

/// Aggregated state for one sandbox instance (C++ `SandboxInfo`).
#[derive(Debug, Clone, Default, PartialEq)]
pub struct SandboxInfo {
    pub runtime_id: String,
    pub sandbox_id: String,
    pub checkpoint_id: String,
    pub port_mappings_json: String,
    pub instance_info: Option<RuntimeInstanceInfo>,
}

#[derive(Default)]
struct Inner {
    sandboxes: HashMap<String, SandboxInfo>,
    in_progress: HashSet<String>,
    pending_deletes: HashSet<String>,
    warm_up: HashMap<String, FunctionRuntime>,
}

#[derive(Default)]
pub struct RuntimeStateManager {
    inner: Mutex<Inner>,
}

impl RuntimeStateManager {
    pub fn new() -> Self {
        Self::default()
    }

    // ── Registration (atomic) ──────────────────────────────────────────────
    pub fn register(&self, info: SandboxInfo) {
        let mut g = self.inner.lock();
        g.sandboxes.insert(info.runtime_id.clone(), info);
    }

    pub fn unregister(&self, runtime_id: &str) {
        let mut g = self.inner.lock();
        g.sandboxes.remove(runtime_id);
        g.in_progress.remove(runtime_id);
        g.pending_deletes.remove(runtime_id);
        g.warm_up.remove(runtime_id);
    }

    // ── Queries ─────────────────────────────────────────────────────────────
    pub fn find(&self, runtime_id: &str) -> Option<SandboxInfo> {
        self.inner.lock().sandboxes.get(runtime_id).cloned()
    }

    pub fn has_sandbox(&self, runtime_id: &str) -> bool {
        self.inner.lock().sandboxes.contains_key(runtime_id)
    }

    pub fn get_sandbox_id(&self, runtime_id: &str) -> String {
        self.inner
            .lock()
            .sandboxes
            .get(runtime_id)
            .map(|s| s.sandbox_id.clone())
            .unwrap_or_default()
    }

    pub fn get_checkpoint_id(&self, runtime_id: &str) -> String {
        self.inner
            .lock()
            .sandboxes
            .get(runtime_id)
            .map(|s| s.checkpoint_id.clone())
            .unwrap_or_default()
    }

    pub fn get_port_mappings_json(&self, runtime_id: &str) -> String {
        self.inner
            .lock()
            .sandboxes
            .get(runtime_id)
            .map(|s| s.port_mappings_json.clone())
            .unwrap_or_default()
    }

    /// Reverse lookup: runtime_id owning a given sandbox_id (empty if none).
    pub fn find_runtime_id_by_sandbox_id(&self, sandbox_id: &str) -> String {
        self.inner
            .lock()
            .sandboxes
            .values()
            .find(|s| s.sandbox_id == sandbox_id)
            .map(|s| s.runtime_id.clone())
            .unwrap_or_default()
    }

    pub fn all_sandboxes(&self) -> HashMap<String, SandboxInfo> {
        self.inner.lock().sandboxes.clone()
    }

    pub fn active_count(&self) -> usize {
        self.inner.lock().sandboxes.len()
    }

    // ── Partial updates ───────────────────────────────────────────────────────
    pub fn update_sandbox_id(&self, runtime_id: &str, sandbox_id: &str) {
        if let Some(s) = self.inner.lock().sandboxes.get_mut(runtime_id) {
            s.sandbox_id = sandbox_id.to_string();
        }
    }

    pub fn update_checkpoint(&self, runtime_id: &str, checkpoint_id: &str) {
        if let Some(s) = self.inner.lock().sandboxes.get_mut(runtime_id) {
            s.checkpoint_id = checkpoint_id.to_string();
        }
    }

    pub fn clear_checkpoint_id(&self, runtime_id: &str) {
        if let Some(s) = self.inner.lock().sandboxes.get_mut(runtime_id) {
            s.checkpoint_id.clear();
        }
    }

    pub fn update_port_mappings(&self, runtime_id: &str, port_mappings_json: &str) {
        if let Some(s) = self.inner.lock().sandboxes.get_mut(runtime_id) {
            s.port_mappings_json = port_mappings_json.to_string();
        }
    }

    // ── In-progress start tracking (dedup) ─────────────────────────────────────
    pub fn mark_start_in_progress(&self, runtime_id: &str) {
        self.inner.lock().in_progress.insert(runtime_id.to_string());
    }

    pub fn mark_start_done(&self, runtime_id: &str) {
        self.inner.lock().in_progress.remove(runtime_id);
    }

    pub fn is_start_in_progress(&self, runtime_id: &str) -> bool {
        self.inner.lock().in_progress.contains(runtime_id)
    }

    // ── Pending-delete tracking (Stop arriving mid-Start) ──────────────────────
    pub fn mark_pending_delete(&self, runtime_id: &str) {
        self.inner
            .lock()
            .pending_deletes
            .insert(runtime_id.to_string());
    }

    pub fn clear_pending_delete(&self, runtime_id: &str) {
        self.inner.lock().pending_deletes.remove(runtime_id);
    }

    pub fn is_pending_delete(&self, runtime_id: &str) -> bool {
        self.inner.lock().pending_deletes.contains(runtime_id)
    }

    // ── Warm-up state (orthogonal to active sandboxes) ─────────────────────────
    pub fn register_warm_up(&self, runtime_id: &str, proto: FunctionRuntime) {
        self.inner
            .lock()
            .warm_up
            .insert(runtime_id.to_string(), proto);
    }

    pub fn unregister_warm_up(&self, runtime_id: &str) {
        self.inner.lock().warm_up.remove(runtime_id);
    }

    pub fn is_warm_up(&self, runtime_id: &str) -> bool {
        self.inner.lock().warm_up.contains_key(runtime_id)
    }

    pub fn get_warm_up(&self, runtime_id: &str) -> Option<FunctionRuntime> {
        self.inner.lock().warm_up.get(runtime_id).cloned()
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn info(id: &str) -> SandboxInfo {
        SandboxInfo {
            runtime_id: id.to_string(),
            sandbox_id: format!("sb-{id}"),
            ..Default::default()
        }
    }

    #[test]
    fn register_find_unregister() {
        let m = RuntimeStateManager::new();
        m.register(info("r1"));
        assert!(m.has_sandbox("r1"));
        assert_eq!(m.get_sandbox_id("r1"), "sb-r1");
        assert_eq!(m.find_runtime_id_by_sandbox_id("sb-r1"), "r1");
        m.unregister("r1");
        assert!(!m.has_sandbox("r1"));
        assert_eq!(m.find_runtime_id_by_sandbox_id("sb-r1"), "");
    }

    #[test]
    fn unregister_clears_all_associated_state_atomically() {
        let m = RuntimeStateManager::new();
        m.register(info("r1"));
        m.mark_start_in_progress("r1");
        m.mark_pending_delete("r1");
        m.unregister("r1");
        assert!(!m.is_start_in_progress("r1"));
        assert!(!m.is_pending_delete("r1"));
        assert!(!m.has_sandbox("r1"));
    }

    #[test]
    fn partial_updates() {
        let m = RuntimeStateManager::new();
        m.register(info("r1"));
        m.update_sandbox_id("r1", "sb-new");
        m.update_checkpoint("r1", "ckpt-1");
        m.update_port_mappings("r1", "tcp:40001:8080");
        let s = m.find("r1").unwrap();
        assert_eq!(s.sandbox_id, "sb-new");
        assert_eq!(s.checkpoint_id, "ckpt-1");
        assert_eq!(s.port_mappings_json, "tcp:40001:8080");
        m.clear_checkpoint_id("r1");
        assert_eq!(m.get_checkpoint_id("r1"), "");
    }

    #[test]
    fn in_progress_and_pending_delete_flags() {
        let m = RuntimeStateManager::new();
        assert!(!m.is_start_in_progress("r1"));
        m.mark_start_in_progress("r1");
        assert!(m.is_start_in_progress("r1"));
        m.mark_pending_delete("r1");
        assert!(m.is_pending_delete("r1"));
        m.mark_start_done("r1");
        m.clear_pending_delete("r1");
        assert!(!m.is_start_in_progress("r1"));
        assert!(!m.is_pending_delete("r1"));
    }

    #[test]
    fn warm_up_is_orthogonal() {
        let m = RuntimeStateManager::new();
        m.register_warm_up("w1", FunctionRuntime::default());
        assert!(m.is_warm_up("w1"));
        assert!(!m.has_sandbox("w1"));
        assert!(m.get_warm_up("w1").is_some());
        m.unregister_warm_up("w1");
        assert!(!m.is_warm_up("w1"));
    }
}
